File: DynamicPrograming/php/canConcat.php
<?php

declare(strict_types=1);

/** @param string[] $words */
function canConcat(string $s, array $words): bool
{
    $memo = [];

    $dfs = function (int $index) use (&$dfs, &$memo, $s, $words): bool {
        if ($index === strlen($s)) {
            return true;
        }
        if (array_key_exists($index, $memo)) {
            return $memo[$index];
        }

        foreach ($words as $word) {
            $length = strlen($word);
            if (substr($s, $index, $length) === $word && $dfs($index + $length)) {
                return $memo[$index] = true;
            }
        }

        return $memo[$index] = false;
    };

    return $dfs(0);
}

File: DynamicPrograming/go/can_concat.go
package dynamicprogramming

// CanConcat reports whether s can be formed by concatenating words.
func CanConcat(s string, words []string) bool {
	memo := make(map[int]bool)

	var dfs func(int) bool
	dfs = func(i int) bool {
		if i == len(s) {
			return true
		}
		if val, ok := memo[i]; ok {
			return val
		}
		for _, word := range words {
			if len(s)-i >= len(word) && s[i:i+len(word)] == word {
				if dfs(i + len(word)) {
					memo[i] = true
					return true
				}
			}
		}
		memo[i] = false
		return false
	}

	return dfs(0)
}
